vllm/compilation/qk_norm_rope_fusion.py

```diff
@@ -0,0 +1,278 @@
+# SPDX-License-Identifier: Apache-2.0
+# SPDX-FileCopyrightText: Copyright contributors to the vLLM project
+"""Pattern-based fusion pass for Q/K RMSNorm + RoPE -> fused_qk_norm_rope."""
+
+import operator
+
+import torch
+import torch._inductor.pattern_matcher as pm
+from torch import fx
+from torch._higher_order_ops.auto_functionalize import auto_functionalized
+from torch._inductor.pattern_matcher import PatternMatcherPass
+
+from vllm.attention import Attention
+from vllm.config import VllmConfig, get_layers_from_vllm_config
+from vllm.logger import init_logger
+from vllm.platforms import current_platform
+
+from .fusion import empty_bf16, empty_i64
+from .inductor_pass import enable_fake_mode
+from .vllm_inductor_pass import VllmInductorPass, VllmPatternMatcherPass
+
+logger = init_logger(__name__)
+
+
+# Ops used in the pattern (assume kernels are built and available)
+RMS_OP = torch.ops._C.rms_norm.default
+ROPE_OPS: list[torch._ops.OpOverload] = [
+    torch.ops._C.rotary_embedding.default,
+    # torch.ops.vllm.flashinfer_rotary_embedding.default,
+]
+FUSED_QK_ROPE_OP = torch.ops._C.fused_qk_norm_rope.default
+SPLIT_SIZES_OP = torch.ops.aten.split_with_sizes.default
+RESHAPE_OP = torch.ops.aten.reshape.default
+EMPTY_LIKE_OP = torch.ops.aten.empty_like.default
+VIEW_OP = torch.ops.aten.view.default
+CONTIGUOUS_OP = torch.ops.aten.contiguous.default
+
+
+class QkNormRopePattern:
+    """
+    Match the unfused sequence in attention blocks and replace with the fused op.
+
+    Unfused (conceptually):
+      q, k, v = split(qkv, [qsz, kvsz, kvsz], -1)
+      qh = reshape(q, [-1, num_heads, head_dim])
+      kh = reshape(k, [-1, num_kv_heads, head_dim])
+      qn = rms_norm(qh, q_weight, eps)
+      kn = rms_norm(kh, k_weight, eps)
+      qf = reshape(qn, [-1, num_heads * head_dim])
+      kf = reshape(kn, [-1, num_kv_heads * head_dim])
+      qf, kf = rotary_embedding(positions, qf, kf, head_dim, cos_sin_cache, is_neox)
+      return qf, kf, v
+
+    Fused replacement:
+      fused_qk_norm_rope(qkv, num_heads, num_kv_heads, num_kv_heads, head_dim,
+                         eps, q_weight, k_weight, cos_sin_cache, is_neox,
+                         positions.view(-1))
+      return split(qkv, [qsz, kvsz, kvsz], -1)
+    """
+
+    def __init__(
+        self,
+        layer: Attention,
+        eps: float,
+        rope_op: torch._ops.OpOverload,
+        is_neox: bool,
+    ) -> None:
+        self.layer = layer
```

Comment:
Are you just using layer for these sizes? If yes, don't save the layer object on the pattern object, just extract the size properties